### PR TITLE
Batch Delete server changes

### PR DESCRIPTION
**Pull request by piyujai**
## Summary
Ambry batch delete support - server changes

## Description
Existing delete operation handles one blob in a single request, the batch delete change introduces handling on server side of a BatchDeleteRequest received from frontend OperationController. At the store level, this includes deleting a list of blobs in a partition and handling the aggregated response in batchDelete util. This is needed for deletion of composite blobs, along with significant performance improvement on large sized composite blobs.

## Testing
Unit tests done.
Integration tests done.

### CODE REVIEW

**Review by manbearpig1996 — COMMENTED** [review]
What does current delete api do for batch request ?
What does new api do differently ?

**Comment by piyujai** [comment]
> What does current delete api do for batch request ? What does new api do differently ?

Current delete api has a list of blobs in params, but handles it in a All or Nothing way, since it raises exception on first error in  the blob list. Also, all current usages of the existing delete operation is in the form of a singleton list itself, so no partial failure handling is required as well.
 In Batch Delete store util, we are handling a list of blobs for all possible cases (all_success, partial_failure, full_failure) by returning an aggregated response.

File: ambry-api/src/main/java/com/github/ambry/protocol/RequestAPI.java [hunk 1]
```diff
@@ -55,6 +55,15 @@ public interface RequestAPI {
    */
   void handleDeleteRequest(NetworkRequest request) throws IOException, InterruptedException;
 
+  /**
+   * Deletes a list of blobs from the store.
```
Comment (manbearpig1996) [inline ambry-api/src/main/java/com/github/ambry/protocol/RequestAPI.java:59]: Can you add more info in your comment about what this api does, like handling partial complete and full delete scenario.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 2]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
             metrics.deleteBlobTotalTimeInMs, null, null, totalTimeSpent));
   }
 
+  /**
+   * Handles a {@link BatchDeleteRequest} and sends out the response via the {@link RequestResponseChannel}.
```
Comment (manbearpig1996) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:523]: Can you add more info in your comment about what this api does, like handling partial complete and full delete scenario.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 3]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+   * @param request the {@link NetworkRequest} that contains the {@link BatchDeleteRequest}.
+   * @throws IOException if there is an I/O error reading from the {@link NetworkRequest}.
+   * @throws InterruptedException if the operation is interrupted.
+   */
+  @Override
+  public void handleBatchDeleteRequest(NetworkRequest request) throws IOException, InterruptedException {
```
Comment (manbearpig1996) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:529]: Can you add comment on cod where you are handling partial error scenarios

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 4]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+    logger.info("Batch delete request received.");
```
Comment (manbearpig1996) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:530]: Can we add trace logs in the condition where you are handling partial error scenarios ?

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 5]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+    BatchDeleteRequest batchDeleteRequest;
+    if (request instanceof LocalChannelRequest) {
+      // This is a case where handleDeleteRequest is called when frontends are talking to Azure. In this case, this method
+      // is called by request handler threads running within the frontend router itself. So, the request can be directly
```
Comment (manbearpig1996) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:534]: Can you add more comments in code for understanding purpose ?

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java [hunk 6]
```diff
@@ -772,6 +775,183 @@ public void delete(List<MessageInfo> infosToDelete) throws StoreException {
     }
   }
 
+  @Override
+  public StoreBatchDeleteInfo batchDelete(List<MessageInfo> infosToDelete) throws StoreException {
```
Comment (manbearpig1996) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:779]: Can you add comments here too with trace logs ?

**Review by DevenWalia — COMMENTED** [review]
Review phase 1 done.
Pending - AmbryRequests, BlobStore, BlobStoreTest

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 7]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+        batchDeleteRequest.accept(metricsUpdater);
+        for (BlobDeleteStatus blobDeleteStatus: partitionResponseInfo.getBlobsDeleteStatus()){
+          // Set operation level error to Bad_Request if any of the blobs fails
+          ServerErrorCode error = blobDeleteStatus.getStatus();
+          if (error != ServerErrorCode.No_Error) {
+            // Handling for control test when request type is disabled + setting default to Bad_Request for any error.
+            operationServerErrorCode = error==ServerErrorCode.Temporarily_Disabled?ServerErrorCode.Temporarily_Disabled:ServerErrorCode.Bad_Request;
+            boolean logInErrorLevel = false;
+            if (error == ServerErrorCode.Blob_Not_Found) {
+              metrics.idNotFoundError.inc();
```
Comment (skandasa23) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:609]: Can we also please add one more kind of metrics that has counters at BatchDelete level? Will be useful to understand how the batchdelete is trending.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 8]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+              logger.trace("Store exception on a delete within a batch delete operation with error code {} for request {}", error, batchDeleteRequest);
+            }
+          }
+        }
+        partitionResponseInfoList.add(partitionResponseInfo);
+
+      }
+    }
+    response = new BatchDeleteResponse(batchDeleteRequest.getCorrelationId(), batchDeleteRequest.getClientId(), partitionResponseInfoList, operationServerErrorCode);
+    logger.info(response.toString());
```
Comment (skandasa23) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:632]: This might not be needed as well.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 9]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+    requestResponseChannel.sendResponse(response, request,
+    new ServerNetworkResponseMetrics(metrics.deleteBlobResponseQueueTimeInMs, metrics.deleteBlobSendTimeInMs,
```
Comment (skandasa23) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:634]: If you are planning to add metrics at batchdelete level, think its more appropriate to use them here?

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java [hunk 10]
```diff
@@ -772,6 +775,183 @@ public void delete(List<MessageInfo> infosToDelete) throws StoreException {
...
+      List<IndexValue> originalPuts = new ArrayList<>();
+      List<Short> lifeVersions = new ArrayList<>();
+      Offset indexEndOffsetBeforeCheck = index.getCurrentEndOffset();
+      filteredInfos = new ArrayList<>();
+      for (MessageInfo info : infosToDelete) {
+        try {
+          IndexValue value =
+              index.findKey(info.getStoreKey(), new FileSpan(index.getStartOffset(), indexEndOffsetBeforeCheck));
+          if (value == null) {
+            throw new StoreException(
```
Comment (skandasa23) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:798]: Seems the infoToDelete are iterated twice

    validate and filter only the qualified Ids
    Issue delete for the ids identified via#1

Iteration(1) seems to be throwing error when an id is not present in index..etc, think we need to handle this exception local to the function and ensure the batchDelete operation in attempted in fullest.

**Review by skandasa23 — CHANGES_REQUESTED** [review]

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java (hunk 6 above)
Comment (DevenWalia) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:779]: Can we look at breaking this function in smaller pieces?
Currently it stands at 180+ loc.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java [hunk 11]
```diff
@@ -772,6 +775,183 @@ public void delete(List<MessageInfo> infosToDelete) throws StoreException {
...
+    final Timer.Context context = metrics.batchDeleteResponse.time();
+    int i = 0;
+    List<MessageInfo> filteredInfos = null;
+    try {
+      List<IndexValue> indexValuesPriorToDelete = new ArrayList<>();
+      List<IndexValue> originalPuts = new ArrayList<>();
+      List<Short> lifeVersions = new ArrayList<>();
+      Offset indexEndOffsetBeforeCheck = index.getCurrentEndOffset();
+      filteredInfos = new ArrayList<>();
+      for (MessageInfo info : infosToDelete) {
```
Comment (DevenWalia) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:793]: This corresponds to `infoList` which further corresponds to `convertedStoreKeys` in calling method in `AmbryRequests` class.
Do we have a upper bound on how big this list can be? 

Per design, there needs to be an upper bound to these batches.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java [hunk 12]
```diff
@@ -772,6 +775,183 @@ public void delete(List<MessageInfo> infosToDelete) throws StoreException {
...
+              index.findKey(info.getStoreKey(), new FileSpan(index.getStartOffset(), indexEndOffsetBeforeCheck));
+          if (value == null) {
+            throw new StoreException(
+                "Cannot delete id " + info.getStoreKey() + " because it is not present in the index",
+                StoreErrorCodes.ID_Not_Found);
+          }
+          if (!info.getStoreKey().isAccountContainerMatch(value.getAccountId(), value.getContainerId())) {
+            if (config.storeValidateAuthorization) {
+              throw new StoreException(
+                  "DELETE authorization failure. Key: " + info.getStoreKey() + "Actually accountId: "
```
Comment (DevenWalia) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:805]: nit: Better to create the error string as a separate statement.

```
string error = "DELETE authorization failure. Key: " + info.getStoreKey() + "Actually accountId: " + value.getAccountId() + "Actually containerId: " + value.getContainerId();

if (config.storeValidateAuthorization) {
    throw new StoreException(error);
} else {
    logger.warn(error);
}
```

That way the error strings stay consistent & can evolve consistently across both places.

File: ambry-store/src/main/java/com/github/ambry/store/BlobStore.java [hunk 13]
```diff
@@ -772,6 +775,183 @@ public void delete(List<MessageInfo> infosToDelete) throws StoreException {
...
+          }
+          infosToDelete = filteredInfos;
+          if (infosToDelete.isEmpty()){
+            logger.trace("Batch Delete completely failed since all blobs failed");
+            return storeBatchDeleteInfo;
+          }
+        }
+        List<InputStream> inputStreams = new ArrayList<>(infosToDelete.size());
+        List<MessageInfo> updatedInfos = new ArrayList<>(infosToDelete.size());
+        i = 0;
```
Comment (DevenWalia) [inline ambry-store/src/main/java/com/github/ambry/store/BlobStore.java:909]: How are we using `i` ?
It was being incremented in previous blocks, then set to 0 here and its being incremented again in next blocks but we don't seem to use the value anywhere?

**Review by DevenWalia — CHANGES_REQUESTED** [review]
Reviewed AmbryRequests, BlobStore changes

High level comments 
1. Break functions into smaller chunks. The 180+ loc BlobStore::BatchDelete function is hard to parse.

File: ambry-api/src/main/java/com/github/ambry/store/MessageErrorInfo.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+public class MessageErrorInfo {
```
[the remark this answers is not in the document]
Comment (piyujai) [reply ambry-api/src/main/java/com/github/ambry/store/MessageErrorInfo.java:16]: @DevenWalia We will need this class to be public since it is being accessed outside the com.github.ambry.store package.
Agreed on the lombok annotations, will look to including this new dependency and integrate it atleast for the new classes.

[the remark this answers is not in the document]
Comment (DevenWalia) [reply ambry-api/src/main/java/com/github/ambry/store/MessageErrorInfo.java:16]: Yea, wherever possible, try to use package-private access for new classes.
There might be a whole lot of places where this cannot be done due to the earlier way of packaging classes, which is fine.

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 15]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+    boolean operationSuccessful = true;
+    ServerErrorCode operationServerErrorCode = ServerErrorCode.No_Error;
+    List<BatchDeletePartitionResponseInfo> partitionResponseInfoList = new ArrayList<>();
+    BatchDeleteResponse response = new BatchDeleteResponse(batchDeleteRequest.getCorrelationId(), batchDeleteRequest.getClientId(), new ArrayList<>(), ServerErrorCode.No_Error);
+    for (BatchDeletePartitionRequestInfo batchDeletePartitionRequestInfo : batchDeleteRequest.getPartitionRequestInfoList()) {
+      BatchDeletePartitionResponseInfo partitionResponseInfo = null;
+      try {
+        List<StoreKey> convertedStoreKeys = getConvertedStoreKeys(batchDeletePartitionRequestInfo.getBlobIds());
+        ServerErrorCode error =
+            validateRequest(batchDeletePartitionRequestInfo.getPartition(), RequestOrResponseType.BatchDeleteRequest,
```
Comment (mudit-saxena) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:552]: This method is name confusing. It is validating partition and whether PUT request is going to READ only partition. Also, FYI @piyujai it doesn't use `skipPartitionAndDiskAvailableCheck` in case you are relying on this check for disk and partition availability

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 16]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+        } else {
+          List<MessageInfo> infoList = new ArrayList<>();
+          for (StoreKey convertedStoreKey : convertedStoreKeys) {
+            BlobId convertedBlobId = (BlobId) convertedStoreKey;
+            infoList.add(new MessageInfo.Builder(convertedBlobId, -1, convertedBlobId.getAccountId(),
+                convertedBlobId.getContainerId(), batchDeleteRequest.getDeletionTimeInMs()).isDeleted(true)
+                .lifeVersion(MessageInfo.LIFE_VERSION_FROM_FRONTEND)
+                .build());
+          }
+          Store storeToDelete = storeManager.getStore(batchDeletePartitionRequestInfo.getPartition());
```
Comment (mudit-saxena) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:570]: `storeToDelete` gives a signal that we are deleting this store. Maybe rename this to something like `targetStore`?

File: ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java [hunk 17]
```diff
@@ -508,6 +519,122 @@ public void handleDeleteRequest(NetworkRequest request) throws IOException, Inte
...
+            for (StoreKey convertedStoreKey : convertedStoreKeys) {
+              notification.onBlobReplicaDeleted(currentNode.getHostname(), currentNode.getPort(),
+                  convertedStoreKey.getID(), BlobReplicaSourceType.PRIMARY);
+            }
+          }
+        }
+      } catch (Exception e) {
+        List<BlobDeleteStatus> blobDeleteStatuses = new ArrayList<>();
+        for (StoreKey storeKey : batchDeletePartitionRequestInfo.getBlobIds()){
+          blobDeleteStatuses.add(new BlobDeleteStatus((BlobId)storeKey, ServerErrorCode.Bad_Request));
```
Comment (mudit-saxena) [inline ambry-server/src/main/java/com/github/ambry/server/AmbryRequests.java:589]: There might be cases of partial failure and success, isn't it?

File: ambry-commons/src/main/java/com/github/ambry/commons/ServerMetrics.java [hunk 18]
```diff
@@ -399,6 +403,11 @@ public ServerMetrics(MetricRegistry registry, Class<?> requestClass, Class<?> se
     deleteBlobProcessingTimeInMs = registry.histogram(MetricRegistry.name(requestClass, "DeleteBlobProcessingTime"));
     deleteBlobResponseQueueTimeInMs =
         registry.histogram(MetricRegistry.name(requestClass, "DeleteBlobResponseQueueTime"));
+    batchDeleteBlobRequestQueueTimeInMs =
+        registry.histogram(MetricRegistry.name(requestClass, "BatchDeleteBlobRequestQueueTime"));
```
[the remark this answers is not in the document]
Comment (piyujai) [reply ambry-commons/src/main/java/com/github/ambry/commons/ServerMetrics.java:407]: Possible to elaborate? @mudit-saxena

[the remark this answers is not in the document]
Comment (mudit-saxena) [reply ambry-commons/src/main/java/com/github/ambry/commons/ServerMetrics.java:423]: Should be ok, having a prefix helps when looking at the metric

**Review by mudit-saxena — APPROVED** [review]
GTG

**Review by manbearpig1996 — APPROVED** [review]

**Review by DevenAhluwalia — APPROVED** [review]
🚢
